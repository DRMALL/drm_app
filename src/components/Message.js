import React, { Component } from 'react'
import { View, Text, Image, TouchableOpacity, ScrollView } from 'react-native'
import { primaryColor, mainColor, backgroundColor, subTitleColor } from '../common/constants'
import { messageText, allSetAsRead, inTheEnd } from '../common/strings'
import { message } from '../styles'
import { messagesList } from '../utils/virtualData'

const gobackWhiteIcon = require('../images/navigation_icons/goback_white.png')

export default class Message extends Component {
  static navigationOptions = ({ navigation })=> ({
    headerStyle: {
      backgroundColor: primaryColor,
    },
    headerTitle: <Text style={{ fontSize: 20, color: '#FFF', alignSelf: 'center' }} >{messageText}</Text>,
    headerLeft: <TouchableOpacity style={{paddingLeft: 20}} onPress={() => navigation.goBack()}>
      <Image source={gobackWhiteIcon}/>
    </TouchableOpacity>,
    headerRight: <TouchableOpacity style={{paddingRight: 20}} onPress={() => alert('已设为已读')}>
      <Text style={{fontSize: 14, color: mainColor}}>{allSetAsRead}</Text>
    </TouchableOpacity>,
  })
  render() {
    return (
      <ScrollView style={message.scrollView}>
        { messagesList.map((msgItem, i)=> <MessageItem key={i} msgItem={msgItem} />) }
        <Text style={message.endText}>{inTheEnd}</Text>
      </ScrollView>
    )
  }
}

const MessageItem = props => {
  return (
    <View style={message.itemView}>
      <View style={props.msgItem.read ? message.empty : message.redDot} />
      <View style={message.textPart}>
        <View style={message.topLine}>
          <Text style={message.textTitle}>{props.msgItem.title}</Text>
          <Text style={message.textTime}>{props.msgItem.publish_time}</Text>
        </View>
        <Text style={message.textAbstract}>{props.msgItem.abstract}</Text>
        <Text style={message.textState}>{'状态：' + props.msgItem.state}</Text>
      </View>
    </View>
  )
}

