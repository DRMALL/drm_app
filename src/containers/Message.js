import React, { Component } from 'react'
import { View, Text, Image, TouchableOpacity, ScrollView, Alert } from 'react-native'
import Loading from '../components/units/Loading'
import { primaryColor, mainColor } from '../common/constants'
import { messageText, 
        allSetAsRead, 
        inTheEnd, 
        tokenKey, 
} from '../common/strings'
import { message } from '../styles'
import { checkToken } from '../utils/handleToken'
import { getPort, postPort } from '../utils/fetchMethod'
import { getNotices, setAllNoticesRead } from '../apis'

import MessageItem from '../components/message/MessageItem'
import store from '../utils/store'
import messageAC from '../actions/messageAC'

const gobackWhiteIcon = require('../images/navigation_icons/goback_white.png')

export default class Message extends Component {
  static navigationOptions = ({ navigation })=> ({
    headerStyle: {
      backgroundColor: primaryColor,
    },
    headerTitle: <Text style={{ fontSize: 20, color: '#FFF', alignSelf: 'center' }} >{messageText}</Text>,
    headerLeft: <TouchableOpacity style={{padding: 10, paddingLeft: 20}} onPress={() => navigation.goBack()}>
      <Image source={gobackWhiteIcon}/>
    </TouchableOpacity>,
    headerRight: <TouchableOpacity style={{padding: 10, paddingRight: 20}} onPress={()=> navigation.state.params.setAllRead()}>
      <Text style={{fontSize: 14, color: mainColor}}>{allSetAsRead}</Text>
    </TouchableOpacity>,
  });

  constructor(props) {
    super(props)
    this.state = store.getState().message
  }

  componentDidMount() {
    this.props.navigation.setParams({  
      setAllRead: () => {
        this.postNoticeRead()
      }, 
    })
    this.getNotices()
  }

  componentWillMount() {
    this.unsubscribe = store.subscribe( ()=> this.setState(store.getState().message) )
  }

  componentWillUnmount(){
    this.unsubscribe()
  }

  postNoticeRead() {
    checkToken(tokenKey)
    .then(async token => {
      let res = await postPort(`${setAllNoticesRead}?token=${token}`)
      if(!res) {
        Alert.alert('❌错误', 'Internal Server Error',
          [ {text: 'OK', onPress: () => 'OK'}, ],
          { cancelable: false }
        )
      } else if(res.code == 201) {
        Alert.alert('通知', '全部已设置成已读',
          [ {text: 'OK', onPress: () => 'OK'}, ],
          { cancelable: false }
        )
        this.getNotices()
      } else {
        Alert.alert('❌错误', JSON.stringify(res.message),
          [ {text: 'OK', onPress: () => 'OK'}, ],
          { cancelable: false }
        )
      }
    })
  }

  getNotices() {
    checkToken(tokenKey)
    .then(async token => {
      let res = await getPort(`${getNotices}?token=${token}`)
      if(!res) {
        Alert.alert('❌错误', 'Internal Server Error',
          [ {text: 'OK', onPress: () => 'OK'}, ],
          { cancelable: false }
        )
      } else if(res.code == 200) {
        messageAC.getAll(res.data)
      } else {
        Alert.alert('❌错误', JSON.stringify(res.message),
          [ {text: 'OK', onPress: () => 'OK'}, ],
          { cancelable: false }
        )
      }
    })
  }

  render() {
    let { navigation } = this.props
      , { noticeData } = this.state

    if(!noticeData) return <Loading animating={!noticeData ? true : false}/>
    noticeData.sort((a, b)=> {
      return new Date(b.order.time) - new Date(a.order.time)
    })
    return (
      <ScrollView style={message.scrollView}>
        { noticeData.map((msgItem, i)=> <MessageItem key={i} msgItem={msgItem} navigation={navigation}/>) }
        <Text style={message.endText}>{inTheEnd}</Text>
      </ScrollView>
    )
  }
}

