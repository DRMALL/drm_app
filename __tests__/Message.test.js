import 'react-native'
import React from 'react'
import Message from '../src/containers/Message'
import renderer from 'react-test-renderer';

test('Message test', () => {
  const navigation = {
    setParams: () => 'ok',
  }
  const tree = renderer.create(
    <Message navigation={ navigation }/>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})