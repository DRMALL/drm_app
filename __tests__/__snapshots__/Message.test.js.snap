// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message test 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "padding": 8,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="large"
        style={
          Object {
            "height": 50,
            "marginTop": 150,
          }
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#7B7B7B",
            "fontSize": 11,
          }
        }
      >
        加载中...
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
